devito/types/tensor.py

```diff
@@ -403,6 +409,20 @@ def tens_func(func1, func2):
     return TensorTimeFunction if f1 or f2 else TensorFunction
 
 
+def shift_x0(shift, shape):
```

Comment:
this is almost a copy-paste. Can you generalize the two `shift_x0` and turn them into a single function? and move it to tools? you could also probably use a better name

Reply:
We have `from devito.finite_differences import Differentiable` in tools.py, so I couldn't call `from devito.finite_differences.tools import shift_x0` inside defferentiable.py due to circular import. Any idea to solve that?

Reply:
Nevermind.. solved!

Reply:
@FabioLuporini what do you think? Name chosen based on [this one](https://github.com/devitocodes/devito/blob/3fb0c067d2b8019302af28078537262d6f908518/devito/passes/clusters/utils.py#L91). Let me know about anymore suggestions.

```python
def make_get_shifted_oringin(shift, shape):
    """
    Returns a callable that calculates a shifted origin for each derivative
    of a given operation derivatives scheme (shape) and a shift object which
    can be a None, a single value or a tuple with the given shape.
    """
    if shift is None:
        return lambda s, d, i, j: None
    if type(shift) is float:
        return lambda s, d, i, j: d + s * d.spacing
    if type(shift) is tuple and np.shape(shift) == shape:
        if len(shape) == 1:
            return lambda s, d, i, j: d + s[j] * d.spacing
        if len(shape) == 2:
            return lambda s, d, i, j: d + s[i][j] * d.spacing
    raise ValueError("shift parameter must be one of the following options: "
                     "None, float or tuple with shape %s." % (shape,))
```